Refs #10055 - Initial model bindings for OSTREE CRUD

parthaa opened the pull request.

ehelms commented:
Tests?

parthaa commented:
Yeah .. Need to add those

jlsherrill commented on db/migrate/20150513034751_add_repository_branches.rb:
```diff
@@ -0,0 +1,26 @@
+class AddRepositoryBranches < ActiveRecord::Migration
+  def up
+    create_table :katello_repository_branches do |t|
+      t.string :branch, :null => false
```
any reason to not use 'name' instead of 'branch' ?

parthaa replied:
Ambivalent about this ..  So that column will have data that looks like "redhat-atomic-host/el7.0/x86_64/base"  or "redhat-atomic-host/el7.0/x86_64/medium" 

Ok with name/path/branch -> you get to choose

daviddavis replied:
I'm not fond of having a `branch_paths` method on Repository if the column name is not `path`. Seems discordant.

parthaa replied:
yeah I didn't want to call that method branch_branches :)

parthaa replied:
@daviddavis one of the suggestions from @jlsherrill and @ehelms was rename repository_branches table to ostree_repository_branches and do the same for the model.. Also the other suggestion was to call the branch attribute 'ref' . Look for ref in https://github.com/projectatomic/rpm-ostree/blob/master/doc/compose-server.md#TreefileKeyDocumentation

parthaa replied:
https://github.com/projectatomic/rpm-ostree/blob/master/doc/treefile-examples/fedora-rawhide-base.json#L2 .. Ok more I see examples the more I am convinced it should be refs and not branch or path attribute name. Going to rename....

ehelms replied:
Based on that example I am inclined to agree with 'ref' for the property name, +1

daviddavis commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -92,6 +93,11 @@ def create
+      if repo_params[:branches]
+        repo.branches = repo_params[:branches].map do |branch|
+          ReposioryBranch.create(:repository => repo, :branch => branch)
+        end
```
You can simplify this:

```ruby
repo_params[:branches].each { |branch| repo.branches.create!(:branch => branch } if repo_params[:branches]
```

daviddavis commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -119,9 +125,22 @@ def sync
+
+        # add the new branches
+        (repo_params[:branches] - @repository.branch_paths).each do |path|
+          ReposioryBranch.create!(:repository => @repository, :branch => path)
```
Probably just simplify this with:

```
@repository.branches.create!(:branch => path)
```

daviddavis commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -119,9 +125,22 @@ def sync
+        end
+
+        # add the new branches
+        (repo_params[:branches] - @repository.branch_paths).each do |path|
```
Why are you using `path` here but every other loop uses `branch`?

daviddavis commented on app/models/katello/repository_branch.rb:
```diff
@@ -0,0 +1,14 @@
+    belongs_to :repository, :class_name => "Katello::Repository", :inverse_of => :branches
+    validates :branch, presence: true, uniqueness: {scope: :repository_id}
+    validates :repository_id, :presence => true
+    validate :ensure_branch_only_for_ostree
```
I think I like `ensure_ostree_repository` or `ensure_ostree_repo` better.

daviddavis commented on app/models/katello/glue/pulp/repo.rb:
```diff
@@ -166,6 +171,13 @@ def generate_distributors
+                      :auto_publish => true }
+          dist = Runcible::Models::OstreeDistributor.new(options)
+          [dist, nodes_distributor]
+
```
Remove me. :smile_cat:

ehelms commented on app/models/katello/repository_branch.rb:
```diff
@@ -0,0 +1,14 @@
+    def ensure_branch_only_for_ostree
+      unless self.repository.ostree?
+        errors.add(:base, _("Branch cannot be created since it does not belong to an RPM OSTree repository."))
+      end
```
After watching a Sandi Metz talk, I have become a fan of using guards in situations like this, e.g.:

```
def ensure_branch_only_for_ostree
  return true if self.repository.ostree?
  errors.add(:base, _("Branch cannot be created since it does not belong to an RPM OSTree repository."))
end
```

ehelms commented on app/models/katello/repository.rb:
```diff
@@ -515,5 +527,11 @@ def ensure_valid_docker_attributes
+    def ensure_has_url_for_ostree
+      if url.blank? && library_instance_id.nil?
+        errors.add(field, N_("cannot be blank. RPM OSTree Repository URL required for syncing from the upstream."))
+      end
```
After watching a Sandi Metz talk, I have become a fan of using guards in situations like this, e.g.:

```
def ensure_branch_only_for_ostree
  return true if url.blank? && library_instance_id.nil?
  errors.add(field, N_("cannot be blank. RPM OSTree Repository URL required for syncing from the upstream."))
end
```

ehelms commented on app/models/katello/glue/pulp/repo.rb:
```diff
@@ -129,6 +129,11 @@ def generate_importer
+        when Repository::OSTREE_TYPE
+          options = {}
+          options[:feed] = self.url if self.respond_to?(:url)
+          options[:branches] = self.branch_paths
```
Could reduce this to:

```
options = {:branches => self.branch_paths}
options[:feed] = self.url if self.respond_to?(:url)
```

daviddavis commented on test/fixtures/models/katello_repositories.yml:
```diff
@@ -209,6 +208,17 @@ iso:
   gpg_key_id:           <%= ActiveRecord::Fixtures.identify(:fedora_gpg_key) %>
   content_view_version_id: <%= ActiveRecord::Fixtures.identify(:library_view_version_1) %>
 
+ostree_cdn:
```
I feel like this ought to contain "rhel 7" in this as ostree_cdn is quite generic.

daviddavis commented:
What happened to using the name `ref` as the os tree branch attribute?

parthaa commented:
> Edit: I'm not saying I like one or the other more. Just curious why you chose name.

So I had a discussion with @jlsherrill  and @jortel on this (check out the #pulp logs if you have em for more detail). Apparently "ref" is a very generic thing for ostree. What pulp expects in the 'branch' field is a specific kind of ref. It expects a branch head. @jortel felt that calling it something generic like ref will confuse the user when things like 'tags' come up in the future. He felt name, even though more generic was clearer than calling it ref and @jlsherrill agreed with that. So went with name. 
BTW I tested this ->
```bash
$ hammer repository create --organization-id=1 --product-id=1 --content-type=ostree --ostree-branches=redhat-atomic-host/el7.0/x86_64/base,redhat-atomic-host/el7.0/x86_64/medium --url=http://cdn.qa.redhat.com/content/htb/rhel/server/7/x86_64/extras/ostree/ --name=ostree-repo
```

And it created the right things. I also changed a couple of things. I moved the branch creation to /katello/app/lib/actions/katello/repository/create.rb because it seemed like a better spot than the action controller. Same with the update.

ehelms commented:
Silly question, if it's a branch head, why not call it that? Or is the
object actually an OSTreeRepsoitoryBranchHead ?
On May 18, 2015 6:48 PM, "Partha Aji" <notifications@github.com> wrote:

> Edit: I'm not saying I like one or the other more. Just curious why you
> chose name.
>
> So I had a discussion with @jlsherrill <https://github.com/jlsherrill>
> and @jortel <https://github.com/jortel> on this (check out the #pulp logs
> if you have em for more detail). Apparently "ref" is a very generic thing
> for ostree. What pulp expects in the 'branch' field is a specific kind of
> ref. It expects a branch head. @jortel <https://github.com/jortel> felt
> that calling it something generic like ref will confuse the user when
> things like 'tags' come up in the future. He felt name, even though more
> generic was clearer than calling it ref and @jlsherrill
> <https://github.com/jlsherrill> agreed with that. So went with name.
> BTW I tested this ->
>
> $ hammer repository create --organization-id=1 --product-id=1 --content-type=ostree --ostree-branches=redhat-atomic-host/el7.0/x86_64/base,redhat-atomic-host/el7.0/x86_64/medium --url=http://cdn.qa.redhat.com/content/htb/rhel/server/7/x86_64/extras/ostree/ --name=ostree-repo
>
> And it created the right things. I also changed a couple of things. I
> moved the branch creation to
> /katello/app/lib/actions/katello/repository/create.rb because it seemed
> like a better spot than the action controller. Same with the update.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Katello/katello/pull/5229#issuecomment-103240141>.
>

parthaa commented:
> Silly question, if it's a branch head, why not call it that? Or is the
> object actually an OSTreeRepsoitoryBranchHead ?

I feel name is a safer option at this time for the following reasons
1) The attribute that we are calling 'name' can store either a path like 'redhat-atomic-host/el7.0/x86_64/base'  or a commit hash or a tag (whenever that is implemented). In rpm ostree I hear its referred to as branch_id. So that rules out 'head'.
2) Apparently 'ref' might imply a branch head or a commit hash, but not necessarily a tag (which is still to be added anyway).

daviddavis commented:
APJ from me.

ehelms commented:
This can't be merged until we have Pulp 2.7 builds in nightly and Runcible is fixed and released, correct?

parthaa commented:
Yeah going to pull this in a separate branch till I get runcible running clean with pulp 2.7 alpha.

parthaa commented:
As stated above moving this code to a branch till runcible is fixed and https://pulp.plan.io/issues/995 gets resolved.

Look at the new PR 
https://github.com/Katello/katello/pull/5240
